### PR TITLE
Add delay compensation to header.stamp

### PR DESCRIPTION
**Pull request by maxpolzin**
This PR adds delay compensation to the header.stamp field if the Optitrack cameras provide it. This allows for more accurate data recording (particularly over unstable connections).

### CODE REVIEW

File: mocap_optitrack_driver/launch/optitrack2.launch.py [hunk 1]
```diff
@@ -61,19 +61,19 @@ def generate_launch_description():
     )
 
     # Make the driver node take the 'activate' transition
-    # driver_activate_trans_event = EmitEvent(
```
Comment (maxpolzin) [inline mocap_optitrack_driver/launch/optitrack2.launch.py:64]: I need my driver to activate itself. hence there is this autostart event actiavatedi n the launch file

File: dependency_repos.repos [hunk 2]
```diff
@@ -1,9 +1,9 @@
 repositories:
   mocap_msgs:
     type: git
-    url: https://github.com/MOCAP4ROS2-Project/mocap_msgs.git
```
Comment (fmrico) [inline dependency_repos.repos:4]: Please, remove this file from your commit. I suppose it is related to the fork you did.

File: mocap_optitrack_driver/config/mocap_optitrack_driver_params.yaml [hunk 3]
```diff
@@ -1,12 +1,12 @@
 mocap_optitrack_driver_node:
   ros__parameters:
     connection_type: "Unicast" # Unicast / Multicast
-    server_address: "198.17.0.120"
```
Comment (fmrico) [inline mocap_optitrack_driver/config/mocap_optitrack_driver_params.yaml:4]: Please, remove this file from your commit. This change is related to your configuration.

File: mocap_optitrack_driver/include/mocap_optitrack_driver/mocap_optitrack_driver.hpp [hunk 4]
```diff
@@ -84,21 +84,22 @@ class OptitrackDriverNode : public mocap_control::ControlledLifecycleNode
   void initParameters();
 
   void process_frame(sFrameOfMocapData * data);
+  NatNetClient * client;
```
Comment (fmrico) [inline mocap_optitrack_driver/include/mocap_optitrack_driver/mocap_optitrack_driver.hpp:87]: Why you made this member public?

File: mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp [hunk 5]
```diff
@@ -97,6 +99,28 @@ void NATNET_CALLCONV process_frame_callback(sFrameOfMocapData * data, void * pUs
   static_cast<OptitrackDriverNode *>(pUserData)->process_frame(data);
 }
 
+std::chrono::nanoseconds OptitrackDriverNode::get_optitrack_system_latency(sFrameOfMocapData * data){
+  const bool bSystemLatencyAvailable = data->CameraMidExposureTimestamp != 0;
+
+  if ( bSystemLatencyAvailable ){
```
Comment (fmrico) [inline mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp:105]: Did you do a `colcon test`? I think the linter will fail here for style. You need an space before the '{'

File: mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp [hunk 6]
```diff
@@ -105,13 +129,19 @@ OptitrackDriverNode::process_frame(sFrameOfMocapData * data)
...
+  header.frame_id = "map";
+
   std::map<int, std::vector<mocap_msgs::msg::Marker>> marker2rb;
 
   // Markers
   if (mocap_markers_pub_->get_subscription_count() > 0) {
     mocap_msgs::msg::Markers msg;
-    msg.header.frame_id = "map";
-    msg.header.stamp = now();
+    msg.header = header;
```
Comment (fmrico) [inline mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp:144]: Maybe you could write here what you wrote in lines 135 and 137.

File: mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp [hunk 7]
```diff
@@ -21,6 +21,8 @@
 #include <vector>
 #include <memory>
 
+#include "std_msgs/msg/header.hpp"
```
Comment (fmrico) [inline mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp:24]: Remove it if we agreed in the next comment

**Comment by fmrico** [comment]
Hi @maxpolzin 

Thank you very much for your contribution. The main idea after my review is to limit your PR to the `get_optitrack_system_latency` method, which is the real (and great!!) contribution of this PR. BTW, where is it used?

Thanks again

File: mocap_optitrack_driver/include/mocap_optitrack_driver/mocap_optitrack_driver.hpp (hunk 4 above)
Comment (maxpolzin) [reply mocap_optitrack_driver/include/mocap_optitrack_driver/mocap_optitrack_driver.hpp]: Left-over from a previous commit, when I didn't understand the NatNetClient properly, yet.

File: mocap_optitrack_driver/config/mocap_optitrack_driver_params.yaml (hunk 3 above)
Comment (maxpolzin) [reply mocap_optitrack_driver/config/mocap_optitrack_driver_params.yaml]: Yes agreed. Sorry for the clutter.

File: mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp (hunk 5 above)
Comment (maxpolzin) [reply mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp]: Yes, I did. 

Start testing: Jul 19 00:18 CEST
----------------------------------------------------------
End testing: Jul 19 00:18 CEST


This is my output from the test.log. It states: There where errors, but unfortunately I don't see what kind of errors.
Do you have any idea what I am missing? I updated the bracket, but I am not sure whether I am lint-compliant now or not.

File: mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp (hunk 7 above)
Comment (maxpolzin) [reply mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp]: Done.

File: README.md [hunk 8]
```diff
@@ -1,3 +1,13 @@
+## Fork of [mocap4ros2_optitrack](https://github.com/MOCAP4ROS2-Project/mocap4ros2_optitrack) used for the [ROSbloX](https://rosblox.github.io/) project.
```
[the remark this answers is not in the document]
Comment (maxpolzin) [reply README.md]: Agreed. Removed.

File: mocap_optitrack_driver/launch/optitrack2.launch.py (hunk 1 above)
Comment (maxpolzin) [reply mocap_optitrack_driver/launch/optitrack2.launch.py]: Done.

File: dependency_repos.repos (hunk 2 above)
Comment (maxpolzin) [reply dependency_repos.repos]: Done.

**Comment by maxpolzin** [comment]
Hi @fmrico, 

Thx for your extremely fast review and your comments. I pretty much agree with all of them and updated the branch to resolve them. Now only the system delay compensation function is added. Please excuse the additional clutter in the original PR.


We are using an Optitrack mocap to characterize our robots in the CREATE Lab at EPFL. We are working on a wide range of mobile robots and manipulators, e.g. one for Robocup@Home (similar to your group, if I am not mistaken). I am focusing on systems which operate in extreme environments, e.g. on/in/under glaciers and ice sheets. To simplify development, we have introduced ROSbloX to quickly composition new systems. The Optitrack Mocap ROSbloX is one of them.

**Comment by fmrico** [comment]
Hi @maxpolzin 

I have found that the tests in CI were not enabled. As far as I remember, there were many style errors in the NATNetSDK, so the test was disabled in CI at any moment. You should have seen the error attached when you run "colcon test":

[stdout_stderr.log](https://github.com/MOCAP4ROS2-Project/mocap4ros2_optitrack/files/12091446/stdout_stderr.log)

Now it is corrected in https://github.com/MOCAP4ROS2-Project/mocap4ros2_optitrack/commit/6db09a8a9048d7fe4b98dd56fcda2d9cad4eca85 . Please, merge/rebase with main to get the real CI status for this PR.

Thanks!!

File: mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp (hunk 5 above)
Comment (fmrico) [reply mocap_optitrack_driver/src/mocap_optitrack_driver/mocap_optitrack_driver.cpp]: See below the comment

**Comment by maxpolzin** [comment]
Hi @fmrico, 
I have just pushed changes to make the PR compliant with the linting. 
I am not sure whether I made the correct changes and run tests properly on my machine. 
Hence, please let me know if the linting tests fail and I should add more changes.

**Comment by maxpolzin** [comment]
@fmrico  Is this PR now copmliant with linting?

**Comment by fmrico** [comment]
Hi @maxpolzin 

The changes are OK for me, but the CI is failing. If you check the [results](https://github.com/MOCAP4ROS2-Project/mocap4ros2_optitrack/actions/runs/5848144127/job/16277922884?pr=21), uncrustify is asking for some changes:

```
--- src/mocap_optitrack_driver/mocap_optitrack_driver.cpp
      +++ src/mocap_optitrack_driver/mocap_optitrack_driver.cpp.uncrustify
      @@ -104 +104 @@
      -  if ( bSystemLatencyAvailable ) {
      +  if (bSystemLatencyAvailable) {
      @@ -107 +107 @@
      -    const double clientLatencyMillisec =  clientLatencySec * 1000.0;
      +    const double clientLatencyMillisec = clientLatencySec * 1000.0;
      @@ -109 +109 @@
      -      client->SecondsSinceHostTimestamp(data->TransmitTimestamp ) * 1000.0;
      +      client->SecondsSinceHostTimestamp(data->TransmitTimestamp) * 1000.0;
      @@ -112,4 +112,5 @@
      -    if(clientLatencyMillisec >= largeLatencyThreshold) {
      -      RCLCPP_WARN_THROTTLE(get_logger(), *this->get_clock(), 500,
      -      "Optitrack system latency >%.0f ms: [Transmission: %.0fms, Total: %.0fms]",
      -      largeLatencyThreshold, transitLatencyMillisec, clientLatencyMillisec);
      +    if (clientLatencyMillisec >= largeLatencyThreshold) {
      +      RCLCPP_WARN_THROTTLE(
      +        get_logger(), *this->get_clock(), 500,
      +        "Optitrack system latency >%.0f ms: [Transmission: %.0fms, Total: %.0fms]",
      +        largeLatencyThreshold, transitLatencyMillisec, clientLatencyMillisec);
```

I think I can merge this PR as soon as you fix this and the CI is green :wink: 

Thanks for your work!!!

**Comment by maxpolzin** [comment]
Thx @fmrico. 

Sorry about the back and forth regarding the linting. I don't have uncrustify properly setup atm and thus made the style changes manually.  Is the installation of uncrustify as simple as installing it to my VS Code or does it require specific settings? If yes, where could I find those?

If I am not mistaken, a maintainer has to trigger the CI for the repo. Could you please re-run the CI and check whether it is now style compliant and succeeding?

**Comment by fmrico** [comment]
Hi @maxpolzin 

All tests passed!! Merging 🚀

I have the standard installation of uncrustify that comes with ROS 2. Maybe that particular version for VS is not working well 

Thanks!!!

**Pull request merged**